### ISSUE TITLE
[socket] [windows] Determining type of completion result

### ISSUE THREAD

**Issue by digitalcora** [erlang/otp#7337]
### Problem

In the current API for `nowait` sockets on Windows, the message received when the completion finishes has the pattern

```erlang
{'$socket', Socket, completion, {CompletionHandle, Result}}
```

where `Result` is the same value that would have been returned by the blocking version of the call.

Because there is a lot of overlap in the return types of the various calls (accept, connect, recv*, send*), it's not possible to determine which operation was used from the message alone — `{error, closed}` could have come from any of them, for example.

The expected usage (which mirrors how `select` works) seems to be that you keep the completion handle returned from the initial `nowait` call and match it with the message, so you should know which operation created the handle and thus how to interpret the result. But the ambiguity in the type of the message itself makes it difficult to write a function or library that handles the `Result` in a generic way, independent of the approach used to track completion handles.

I'm currently working on a [Gleam](https://gleam.run/) API for `socket`, and static typing means receiving a completion message is an awkward two-stage process for the library user: first you receive the completion message with the result as an (opaque) untyped value, then depending on what operation you expected for the given handle, you pass that untyped value into a function that attempts to "cast" it to a specific result type. Also, if you call the wrong "cast" function, it can still incorrectly succeed if the shape of the value is the same, leading to potential bugs.

### Solution

I have no preferred solution here, I'd just like it to be possible to correctly "type" a completion message. This would require a change to the message format, which I assumed could be on the table because Windows `socket` support is still in "pre-release" status. If the message format can't be changed, I don't think there is a way to implement this.

One possibility I can think of:

```erlang
{'$socket', Socket, completion, {CompletionHandle, {Operation, Result}}}
```

Where `Operation` is `accept | connect | send | sendto | ...`. I think the internal `socket` code already keeps track of this in the `completion_tag` that is part of `completion_info`, though this is currently an opaque value.

**Comment by bmk** [comment]
Hi,

Sorry for the excessive delay (I was away when this issue was created, and missed it).

We already have "Operation". Its part of the 'CompletionInfo:

{completion, CompletionInfo} = socket:recv(..., nowait)

Where CompletionInfo = {completion_tag(), completion_handle()}.

This is from the documentation:
"A tag that describes the ongoing  (completion)  operation,  contained in the returned completion_info()."

/BMK

**Comment by digitalcora** [comment]
Thanks @bmk, I wasn't sure if the `completion_tag` _should_ be used as-is, since it's opaque and the possible values are not documented. If it's intended that libraries can rely on this value to distinguish the type of the completion, could it be changed to a non-opaque value with a documented set of atoms?

**Comment by bmk** [comment]
A documentation update for this will be released (= is planned for) 26.2.
